src/librustc_mir/interpret/eval_context.rs

```diff
@@ -557,56 +559,142 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+                StackPopCleanup::Goto { ret, ..  } => {
+                    self.goto_block(ret)?;
+                }
+                StackPopCleanup::None { .. } => {}
             }
-        } else {
-            // Uh, that shouldn't happen... the function did not intend to return
-            throw_ub!(Unreachable)
         }
-        // Jump to new block -- *after* validation so that the spans make more sense.
-        match frame.return_to_block {
-            StackPopCleanup::Goto(block) => {
-                self.goto_block(block)?;
+
+
+        Ok((frame.return_to_block, stack_pop_info))
+    }
+
+    pub(super) fn pop_stack_frame(&mut self, unwinding: bool) -> InterpResult<'tcx> {
+        let (mut cleanup, mut stack_pop_info) = self.pop_stack_frame_internal(unwinding)?;
+
+        // There are two cases where we want to unwind the stack:
+        // * The caller explicitly told us (i.e. we hit a Resume terminator)
+        // * The machine indicated that we've just started unwinding (i.e.
+        // a panic has just occured)
+        if unwinding || stack_pop_info == StackPopInfo::StartUnwinding {
+            trace!("unwinding: starting stack unwind...");
+            // Overwrite our current stack_pop_info, so that the check
+            // below doesn't fail.
+            stack_pop_info = StackPopInfo::Normal;
+            // There are three posible ways that we can exit the loop:
+            // 1) We find an unwind block - we jump to it to allow cleanup
+            // to occur for that frame
+            // 2) pop_stack_frame_internal reports that we're no longer unwinding
+            //    - this means that the panic has been caught, and that execution
+            //    should continue as normal
+            // 3) We pop all of our frames off the stack - this should never happen.
+            while !self.stack.is_empty() {
+                match stack_pop_info {
+                    // We tried to start unwinding while we were already
+                    // unwinding. Note that this **is not** the same thing
+                    // as a double panic, which will be intercepted by
+                    // libcore/libstd before we actually attempt to unwind.
+                    StackPopInfo::StartUnwinding => {
+                        throw_ub_format!("Attempted to start unwinding while already unwinding!");
+                    },
+                    StackPopInfo::StopUnwinding => {
+                        trace!("unwinding: no longer unwinding!");
+                        break;
+                    }
+                    StackPopInfo::Normal => {}
+                }
+
+                match cleanup {
+                    StackPopCleanup::Goto { unwind, .. } if unwind.is_some() => {
+
+                        info!("unwind: found cleanup block {:?}", unwind);
+                        self.goto_block(unwind)?;
+                        break;
+                    },
+                    _ => {}
```

Comment:
The only other case is `StackPopCleanup::None`... can that even happen during unwinding?